File: packages/core/src/slide/normalise.ts
import { CoMotionError } from "../errors.js";
import {
  CONTAINER_ATTRIBUTES,
  SLIDE_PRIMITIVE_TAGS,
  checkSlideCompliance,
  type ComplianceCode,
} from "./format.js";
import { attributeOf, scanDocument, type ScannedAttribute, type ScannedNode } from "./scan.js";

/**
 * Normalisation: wrap bare primitives into containers, and move the
 * identifier, display name, media reference and transform up onto the
 * container (ADR-0012).
 *
 * Two things this module deliberately does NOT do:
 *
 * - It does not re-serialize the document. Every byte outside the spans it
 *   splices is preserved exactly, the same invariant `element-text.ts` was
 *   built around (ADR-0001, ADR-0004: SVG size is the per-turn token cost,
 *   and a reflowed file is a diff nobody can read).
 * - It does not hoist a primitive's native coordinates into the container's
 *   `transform` (軍令). `<g><rect x="0" y="0"/></g>` and
 *   `<g transform="translate(0 0)"><rect/></g>` render identically, but
 *   hoisting is only cheap for `rect`/`image`/`text` — for `line` it means
 *   recomputing `x2/y2`, and for `path` it means rewriting the whole `d`.
 *   A rule that only holds for three primitives is not a rule. So the
 *   container's `transform` is where editing commands write position and
 *   rotation, and the primitive's own coordinates stay its internal shape.
 *
 * That combination is what makes "the demo looks pixel-identical after
 * conversion" true by construction rather than by luck: wrapping an element
 * in a `<g>` with no transform changes no rendered geometry at all.
 *
 * No Node built-in imports, not even transitively — hence `generateId` is
 * injected rather than imported (`../id.js` reaches for Node's crypto module).
 */

export interface NormaliseResult {
  /** The normalised document. Byte-identical to the input when it was already compliant. */
  svg: string;
  /** How many new containers were wrapped this run. 0 means nothing changed. */
  wrapped: number;
  /** Identifiers minted for primitives that had none. */
  generatedIds: string[];
}

export interface NormaliseOptions {
  /**
   * Supplies an identifier for a primitive that carries none. Required, not
   * defaulted: core's own `generateElementId()` sits on Node's crypto module, and
   * importing it here would drag Node into a module the front end has to be
   * able to load.
   */
  generateId: () => string;
}

/** Problems conversion can repair. Everything else is the author's to fix first. */
const REPAIRABLE: ReadonlySet<ComplianceCode> = new Set<ComplianceCode>([
  "bare-primitive",
  "missing-id",
  "primitive-transform",
]);

const XML_WHITESPACE = /[\t\n\r ]/;

interface Splice {
  start: number;
  end: number;
  text: string;
}

/**
 * Normalises one slide. Throws when the slide has a problem conversion
 * refuses to repair — `<script>`, `<foreignObject>`, an unknown tag,
 * duplicate identifiers, broken markup — naming the line, so the author
 * fixes the real problem instead of getting a silently mangled file.
 */
export function normaliseSlideSvg(svg: string, options: NormaliseOptions): NormaliseResult {
  const issues = checkSlideCompliance(svg);
  const blocking = issues.find((issue) => !REPAIRABLE.has(issue.code));
  if (blocking) {
    throw new CoMotionError(
      `投影片不合規（第 ${blocking.line} 行第 ${blocking.column} 欄）：${blocking.message}轉換命令不會替你修這一項。`,
    );
  }
  if (issues.length === 0) {
    return { svg, wrapped: 0, generatedIds: [] };
  }

  const roots = scanDocument(svg);
  const svgRoot = roots.find((element) => element.tag === "svg")!;

  const usedIds = new Set<string>();
  collectIds(roots, usedIds);
  const generatedIds: string[] = [];
  const mintId = (): string => {
    for (let attempt = 0; attempt < 8; attempt++) {
      const id = options.generateId();
      if (!id) throw new CoMotionError("轉換失敗：識別碼產生器回傳空字串");
      if (!usedIds.has(id)) {
        usedIds.add(id);
        generatedIds.push(id);
        return id;
      }
    }
    throw new CoMotionError("轉換失敗：連續 8 次產生的識別碼都與既有元素重複");
  };

  const splices: Splice[] = [];
  let wrapped = 0;

  for (const child of svgRoot.children) {
    if (SLIDE_PRIMITIVE_TAGS.includes(child.tag)) {
      splices.push(wrapPrimitive(svg, child, mintId));
      wrapped++;
      continue;
    }
    if (child.tag !== "g") continue;

    for (const container of containersNeedingWork(child)) {
      if (!attributeOf(container, "id")) {
        splices.push({
          start: container.start + 1 + container.tag.length,
          end: container.start + 1 + container.tag.length,
          text: ` id="${mintId()}"`,
        });
      }
      // A container holding a primitive that carries its own `transform` is
      // repaired by giving every one of its primitives a container of its
      // own — the container therefore becomes a group. One predictable rule
      // beats a special case for "the one primitive that had a transform".
      const primitives = container.children.filter((element) => SLIDE_PRIMITIVE_TAGS.includes(element.tag));
      if (primitives.some((element) => attributeOf(element, "transform"))) {
        for (const primitive of primitives) {
          splices.push(wrapPrimitive(svg, primitive, mintId));
          wrapped++;
        }
      }
    }
  }

  return { svg: applySplices(svg, splices), wrapped, generatedIds };
}

/** Every `<g>` at or below `root`, in document order. */
function containersNeedingWork(root: ScannedNode): ScannedNode[] {
  const containers: ScannedNode[] = [];
  const visit = (element: ScannedNode): void => {
    if (element.tag !== "g") return;
    containers.push(element);
    for (const child of element.children) visit(child);
  };
  visit(root);
  return containers;
}

function collectIds(nodes: readonly ScannedNode[], into: Set<string>): void {
  for (const element of nodes) {
    const id = attributeOf(element, "id");
    if (id) into.add(id.value);
    collectIds(element.children, into);
  }
}

/**
 * Builds the replacement text that wraps one primitive in a container,
 * moving `id`, `data-comot-name`, `data-comot-media` and `transform` up
 * onto it. Every other attribute stays on the primitive, in its original
 * order and spacing, byte for byte.
 */
function wrapPrimitive(svg: string, element: ScannedNode, mintId: () => string): Splice {
  const lifted = new Map<string, string>();
  const removals: Array<{ start: number; end: number }> = [];
  for (const name of CONTAINER_ATTRIBUTES) {
    const attribute = attributeOf(element, name);
    if (!attribute) continue;
    lifted.set(name, attribute.value);
    removals.push({ start: startOfAttributeWithLeadingSpace(svg, element, attribute), end: attribute.end });
  }

  let primitiveText = svg.slice(element.start, element.end);
  for (const removal of [...removals].sort((left, right) => right.start - left.start)) {
    primitiveText =
      primitiveText.slice(0, removal.start - element.start) + primitiveText.slice(removal.end - element.start);
  }

  const indent = indentOfLineContaining(svg, element.start);
  const containerAttrs = CONTAINER_ATTRIBUTES.filter((name) => name === "id" || lifted.has(name))
    .map((name) => `${name}=${quote(name === "id" ? (lifted.get("id") ?? mintId()) : lifted.get(name)!)}`)
    .join(" ");

  return {
    start: element.start,
    end: element.end,
    text:
      `<g ${containerAttrs}>\n` +
      `${indent}  ${primitiveText.split("\n").join(`\n  `)}\n` +
      `${indent}</g>`,
  };
}

/** Quotes an attribute value without altering a single character of it. */
function quote(value: string): string {
  if (!value.includes('"')) return `"${value}"`;
  if (!value.includes("'")) return `'${value}'`;
  throw new CoMotionError(`無法搬移含有兩種引號的屬性值：${value}`);
}

/** Where an attribute's removal should start, so the space in front of it goes too. */
function startOfAttributeWithLeadingSpace(
  svg: string,
  element: ScannedNode,
  attribute: ScannedAttribute,
): number {
  const tagNameEnd = element.start + 1 + element.tag.length;
  let start = attribute.start;
  while (start > tagNameEnd && XML_WHITESPACE.test(svg[start - 1])) start--;
  return start;
}

/** The leading whitespace of the line `offset` sits on, or "" when something else precedes it. */
function indentOfLineContaining(svg: string, offset: number): string {
  const lineStart = svg.lastIndexOf("\n", offset - 1) + 1;
  const prefix = svg.slice(lineStart, offset);
  return /^[\t ]*$/.test(prefix) ? prefix : "";
}

function applySplices(svg: string, splices: readonly Splice[]): string {
  let result = svg;
  for (const splice of [...splices].sort((left, right) => right.start - left.start)) {
    result = result.slice(0, splice.start) + splice.text + result.slice(splice.end);
  }
  return result;
}

File: crates/slidra/src/slide/normalise.rs
//! Normalisation (`slidra convert`), ported from
//! `packages/core/src/slide/normalise.ts` (full file, 234 lines): wraps bare
//! primitives into `<g>` containers, and moves `id`/`data-slidra-name`/
//! `data-slidra-media`/`transform` up onto the container (ADR-0012).
//!
//! Does not re-serialize the document — every byte outside the spans it
//! splices is preserved exactly. Does not hoist a primitive's native
//! coordinates into the container's `transform`.

use crate::errors::{SlidraError, SlidraResult};
use crate::slide::format::{
    CONTAINER_ATTRIBUTES, ComplianceCode, SLIDE_PRIMITIVE_TAGS, check_slide_compliance,
};
use crate::slide::scan::{ScannedAttribute, ScannedNode, attribute_of, scan_document};
use crate::splice::{Splice, apply_splices};
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NormaliseResult {
    /// The normalised document. Byte-identical to the input when it was
    /// already compliant.
    pub svg: String,
    /// How many new containers were wrapped this run. 0 means nothing changed.
    pub wrapped: usize,
    /// Identifiers minted for primitives that had none.
    pub generated_ids: Vec<String>,
}

/// Problems conversion can repair. Everything else is the author's to fix
/// first.
fn is_repairable(code: ComplianceCode) -> bool {
    matches!(
        code,
        ComplianceCode::BarePrimitive
            | ComplianceCode::MissingId
            | ComplianceCode::PrimitiveTransform
    )
}

/// Normalises one slide. Errors when the slide has a problem conversion
/// refuses to repair, naming the line, so the author fixes the real problem
/// instead of getting a silently mangled file.
pub fn normalise_slide_svg(
    svg: &str,
    generate_id: &mut dyn FnMut() -> String,
) -> SlidraResult<NormaliseResult> {
    let issues = check_slide_compliance(svg);
    if let Some(blocking) = issues.iter().find(|issue| !is_repairable(issue.code)) {
        return Err(SlidraError::invalid(format!(
            "投影片不合規（第 {} 行第 {} 欄）：{}轉換命令不會替你修這一項。",
            blocking.line, blocking.column, blocking.message
        )));
    }
    if issues.is_empty() {
        return Ok(NormaliseResult {
            svg: svg.to_string(),
            wrapped: 0,
            generated_ids: Vec::new(),
        });
    }

    let roots = scan_document(svg)?;
    let svg_root = roots
        .iter()
        .find(|element| element.tag == "svg")
        .expect("compliance check guarantees an <svg> root");

    let mut used_ids: HashSet<String> = HashSet::new();
    collect_ids(&roots, &mut used_ids);
    let mut generated_ids: Vec<String> = Vec::new();
    let mut mint_id = |used_ids: &mut HashSet<String>| -> SlidraResult<String> {
        for _ in 0..8 {
            let id = generate_id();
            if id.is_empty() {
                return Err(SlidraError::invalid("轉換失敗：識別碼產生器回傳空字串"));
            }
            if !used_ids.contains(&id) {
                used_ids.insert(id.clone());
                generated_ids.push(id.clone());
                return Ok(id);
            }
        }
        Err(SlidraError::invalid(
            "轉換失敗：連續 8 次產生的識別碼都與既有元素重複",
        ))
    };

    let mut splices: Vec<Splice> = Vec::new();
    let mut wrapped = 0usize;

    for child in &svg_root.children {
        if SLIDE_PRIMITIVE_TAGS.contains(&child.tag.as_str()) {
            splices.push(wrap_primitive(svg, child, &mut mint_id, &mut used_ids)?);
            wrapped += 1;
            continue;
        }
        if child.tag != "g" {
            continue;
        }

        for container in containers_needing_work(child) {
            if attribute_of(container, "id").is_none() {
                let id = mint_id(&mut used_ids)?;
                let insert_at = container.start + 1 + container.tag.len();
                splices.push(Splice {
                    start: crate::text::runs::utf16_offset_to_byte_offset(svg, insert_at),
                    end: crate::text::runs::utf16_offset_to_byte_offset(svg, insert_at),
                    text: format!(" id=\"{id}\""),
                });
            }
            let primitives: Vec<&ScannedNode> = container
                .children
                .iter()
                .filter(|element| SLIDE_PRIMITIVE_TAGS.contains(&element.tag.as_str()))
                .collect();
            if primitives
                .iter()
                .any(|element| attribute_of(element, "transform").is_some())
            {
                for primitive in primitives {
                    splices.push(wrap_primitive(svg, primitive, &mut mint_id, &mut used_ids)?);
                    wrapped += 1;
                }
            }
        }
    }

    Ok(NormaliseResult {
        svg: apply_splices(svg, &splices),
        wrapped,
        generated_ids,
    })
}

/// Every `<g>` at or below `root`, in document order (`root` included when
/// it is itself a `<g>`).
fn containers_needing_work(root: &ScannedNode) -> Vec<&ScannedNode> {
    let mut containers = Vec::new();
    fn visit<'a>(element: &'a ScannedNode, containers: &mut Vec<&'a ScannedNode>) {
        if element.tag != "g" {
            return;
        }
        containers.push(element);
        for child in &element.children {
            visit(child, containers);
        }
    }
    visit(root, &mut containers);
    containers
}

fn collect_ids(nodes: &[ScannedNode], into: &mut HashSet<String>) {
    for element in nodes {
        if let Some(id) = attribute_of(element, "id") {
            into.insert(id.value.clone());
        }
        collect_ids(&element.children, into);
    }
}

/// Builds the replacement text that wraps one primitive in a container,
/// moving `id`, `data-slidra-name`, `data-slidra-media` and `transform` up
/// onto it. Every other attribute stays on the primitive, in its original
/// order and spacing, byte for byte.
fn wrap_primitive(
    svg: &str,
    element: &ScannedNode,
    mint_id: &mut dyn FnMut(&mut HashSet<String>) -> SlidraResult<String>,
    used_ids: &mut HashSet<String>,
) -> SlidraResult<Splice> {
    use crate::text::runs::utf16_offset_to_byte_offset as b;

    let mut lifted: Vec<(String, String)> = Vec::new();
    // Byte offsets, computed directly — see `start_of_attribute_with_leading_space`.
    let mut byte_removals: Vec<(usize, usize)> = Vec::new();
    for &name in CONTAINER_ATTRIBUTES {
        let Some(attribute) = attribute_of(element, name) else {
            continue;
        };
        lifted.push((name.to_string(), attribute.value.clone()));
        byte_removals.push((
            start_of_attribute_with_leading_space(svg, element, attribute),
            b(svg, attribute.end),
        ));
    }

    let element_start_byte = b(svg, element.start);
    let element_end_byte = b(svg, element.end);
    let mut primitive_text = svg[element_start_byte..element_end_byte].to_string();
    byte_removals.sort_by(|a, b| b.0.cmp(&a.0));
    for (start, end) in byte_removals {
        let rel_start = start - element_start_byte;
        let rel_end = end - element_start_byte;
        primitive_text = format!(
            "{}{}",
            &primitive_text[..rel_start],
            &primitive_text[rel_end..]
        );
    }

    let indent = indent_of_line_containing(svg, element_start_byte);
    let get_lifted = |name: &str| -> Option<String> {
        lifted
            .iter()
            .find(|(k, _)| k == name)
            .map(|(_, v)| v.clone())
    };

    let mut container_attrs: Vec<String> = Vec::new();
    for &name in CONTAINER_ATTRIBUTES {
        if name == "id" {
            let value = match get_lifted("id") {
                Some(v) => v,
                None => mint_id(used_ids)?,
            };
            container_attrs.push(format!("id={}", quote(&value)?));
        } else if let Some(value) = get_lifted(name) {
            container_attrs.push(format!("{name}={}", quote(&value)?));
        }
    }

    let indented_primitive = primitive_text.split('\n').collect::<Vec<_>>().join("\n  ");
    let text = format!(
        "<g {}>\n{indent}  {indented_primitive}\n{indent}</g>",
        container_attrs.join(" ")
    );

    Ok(Splice {
        start: element_start_byte,
        end: element_end_byte,
        text,
    })
}

/// Quotes an attribute value without altering a single character of it.
fn quote(value: &str) -> SlidraResult<String> {
    if !value.contains('"') {
        return Ok(format!("\"{value}\""));
    }
    if !value.contains('\'') {
        return Ok(format!("'{value}'"));
    }
    Err(SlidraError::invalid(format!(
        "無法搬移含有兩種引號的屬性值：{value}"
    )))
}

/// Where an attribute's removal should start, so the space in front of it
/// goes too. Returns a Rust byte offset (already converted from `element`/
/// `attribute`'s native UTF-16 offset space).
fn start_of_attribute_with_leading_space(
    svg: &str,
    element: &ScannedNode,
    attribute: &ScannedAttribute,
) -> usize {
    use crate::text::runs::utf16_offset_to_byte_offset as b;
    let tag_name_end_byte = b(svg, element.start) + 1 + element.tag.len();
    let mut start_byte = b(svg, attribute.start);
    while start_byte > tag_name_end_byte {
        let Some(prev) = svg[..start_byte].chars().next_back() else {
            break;
        };
        if prev == '\t' || prev == '\n' || prev == '\r' || prev == ' ' {
            start_byte -= prev.len_utf8();
        } else {
            break;
        }
    }
    // Convert back: since UTF-16 and byte space agree in lockstep only when
    // walking whitespace one char at a time from a byte position we already
    // hold, we instead return byte offsets throughout and let the caller
    // treat `removals` as byte offsets directly (see `wrap_primitive`).
    start_byte
}

/// The leading whitespace of the line `offset_byte` (a Rust byte offset)
/// sits on, or "" when something else precedes it.
fn indent_of_line_containing(svg: &str, offset_byte: usize) -> String {
    let line_start = svg[..offset_byte].rfind('\n').map(|i| i + 1).unwrap_or(0);
    let prefix = &svg[line_start..offset_byte];
    if prefix.chars().all(|c| c == '\t' || c == ' ') {
        prefix.to_string()
    } else {
        String::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn gen_ids(ids: Vec<&'static str>) -> impl FnMut() -> String {
        let mut iter = ids.into_iter();
        move || iter.next().unwrap_or("fallback-id").to_string()
    }

    #[test]
    fn already_compliant_slide_is_byte_identical() {
        let svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\"><g id=\"a\"><rect width=\"1\" height=\"1\"/></g></svg>";
        let mut gen_fn = gen_ids(vec![]);
        let result = normalise_slide_svg(svg, &mut gen_fn).unwrap();
        assert_eq!(result.svg, svg);
        assert_eq!(result.wrapped, 0);
        assert!(result.generated_ids.is_empty());
    }

    #[test]
    fn wraps_bare_primitive_and_mints_id() {
        let svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\">\n  <rect width=\"1\" height=\"1\"/>\n</svg>";
        let mut gen_fn = gen_ids(vec!["gen_fn-1"]);
        let result = normalise_slide_svg(svg, &mut gen_fn).unwrap();
        assert_eq!(result.wrapped, 1);
        assert_eq!(result.generated_ids, vec!["gen_fn-1".to_string()]);
        assert!(result.svg.contains("<g id=\"gen_fn-1\">"));
        assert!(result.svg.contains("<rect width=\"1\" height=\"1\"/>"));
    }

    #[test]
    fn container_missing_id_gets_one_minted() {
        let svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\"><g><rect width=\"1\" height=\"1\"/></g></svg>";
        let mut gen_fn = gen_ids(vec!["gen_fn-2"]);
        let result = normalise_slide_svg(svg, &mut gen_fn).unwrap();
        assert!(result.svg.contains("<g id=\"gen_fn-2\">"));
    }

    #[test]
    fn primitive_transform_inside_container_gets_its_own_wrapper() {
        let svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\"><g id=\"outer\"><rect id=\"r\" transform=\"translate(1 1)\" width=\"1\" height=\"1\"/></g></svg>";
        let mut gen_fn = gen_ids(vec![]);
        let result = normalise_slide_svg(svg, &mut gen_fn).unwrap();
        assert_eq!(result.wrapped, 1);
        assert!(
            result
                .svg
                .contains("<g id=\"r\" transform=\"translate(1 1)\">")
        );
        assert!(!result.svg.contains("<rect id=\"r\" transform"));
    }

    #[test]
    fn blocking_issue_errors_with_line_and_column() {
        let svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 100 100\"><script>evil()</script></svg>";
        let mut gen_fn = gen_ids(vec![]);
        let err = normalise_slide_svg(svg, &mut gen_fn).unwrap_err();
        assert!(err.message().contains("不合規"));
        assert!(err.message().contains("轉換命令不會替你修這一項"));
    }
}
